# Cardiac Patient Monitoring System
## AI & Machine Learning Track 


## Project & Dataset Selection

**Chosen project topic:** *Cardiac Patient Monitoring System*.

**Chosen dataset:** the **UCI Heart Disease dataset** (Cleveland database, UCI repository ID 45). It contains **303 patient records** described by **13 clinical features** (age, sex, chest pain type, resting blood pressure, cholesterol, fasting blood sugar, resting ECG, max heart rate, exercise-induced angina, ST depression, slope, number of major vessels, thalassemia result), with a target column indicating heart disease presence/severity (0 = no disease, 1-4 = increasing severity).

**Why this dataset was chosen:**
- It is a **public, non-identifiable** dataset (no names, dates, or medical-record numbers), which matches the project's out-of-scope constraint on identifiable patient data.
- It is **curriculum-aligned**: it needs only Pandas/NumPy for cleaning, Matplotlib for EDA, and Scikit-learn for classification, feature engineering/pipelines, and clustering/PCA — no deep learning or external infrastructure.
- It is a well-documented, classic classification dataset, so results can be reasoned about and sanity-checked against known findings in the literature.
- Its feature set is small and mostly numeric/ordinal, which keeps the feature-engineering and pipeline work (Milestone M5) and the unsupervised analysis (Milestone M6) meaningful and easy to interpret.

**How it was obtained:** the dataset is loaded programmatically through the `ucimlrepo` Python package, using `fetch_ucirepo(id=45)`, which pulls the official UCI Machine Learning Repository entry for *Heart Disease* directly. This keeps data acquisition reproducible — no manual file downloads are needed to re-run this project.

---
## M1 — Environment, Dataset & Data Preparation

### 1. Load Dataset

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from ucimlrepo import fetch_ucirepo

# Load the UCI Heart Disease dataset
heart_disease = fetch_ucirepo(id=45)

X = heart_disease.data.features
y = heart_disease.data.targets

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (303, 13)
Target shape: (303, 1)


In [ ]:
# Display the first few rows of the features
X.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0


In [ ]:
# show the information about the features
X.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
dtypes: float64(3), int64(10)
memory usage: 30.9 KB


In [ ]:
# show the summary statistics of the features
X.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,299.000000,301.000000
mean,54.438944,0.679868,3.158416,131.689769,246.693069,0.148515,0.990099,149.607261,0.326733,1.039604,1.600660,0.672241,4.734219
std,9.038662,0.467299,0.960126,17.599748,51.776918,0.356198,0.994971,22.875003,0.469794,1.161075,0.616226,0.937438,1.939706
min,29.000000,0.000000,1.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,1.000000,0.000000,3.000000
25%,48.000000,0.000000,3.000000,120.000000,211.000000,0.000000,0.000000,133.500000,0.000000,0.000000,1.000000,0.000000,3.000000
50%,56.000000,1.000000,3.000000,130.000000,241.000000,0.000000,1.000000,153.000000,0.000000,0.800000,2.000000,0.000000,3.000000
75%,61.000000,1.000000,4.000000,140.000000,275.000000,0.000000,2.000000,166.000000,1.000000,1.600000,2.000000,1.000000,7.000000
max,77.000000,1.000000,4.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,3.000000,3.000000,7.000000


In [ ]:
# Check for missing values in the features
X.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
dtype: int64

In [ ]:
# Check for duplicate rows in the features
X.duplicated().sum()

np.int64(0)

In [ ]:
# Check the distribution of the target variable
y.value_counts()

num
0      164
1       55
2       36
3       35
4       13
Name: count, dtype: int64

### 2. Data Dictionary

## Data Dictionary

| Feature    | Meaning                                       
---------------------------------------------- 
| `age`      | Age in years                                  
| `sex`      | Sex                                            
| `cp`       | Chest pain type                                
| `trestbps` | Resting blood pressure                         
| `chol`     | Serum cholesterol                              
| `fbs`      | Fasting blood sugar > 120 mg/dl                
| `restecg`  | Resting electrocardiographic results           
| `thalach`  | Maximum heart rate achieved                    
| `exang`    | Exercise-induced angina                        
| `oldpeak`  | ST depression induced by exercise              
| `slope`    | Slope of peak exercise ST segment              
| `ca`       | Number of major vessels colored by fluoroscopy 
| `thal`     | Thalassemia result                             


In [ ]:
# Display feature names
X.columns.tolist()

['age',
 'sex',
 'cp',
 'trestbps',
 'chol',
 'fbs',
 'restecg',
 'thalach',
 'exang',
 'oldpeak',
 'slope',
 'ca',
 'thal']

In [ ]:
# Display the official UCI variable information
heart_disease.variables

,name,role,type,demographic,description,units,missing_values
0,age,Feature,Integer,Age,NaN,years,no
1,sex,Feature,Categorical,Sex,NaN,NaN,no
2,cp,Feature,Categorical,NaN,NaN,NaN,no
3,trestbps,Feature,Integer,NaN,resting blood pressure (on admission to the ho...,mm Hg,no
4,chol,Feature,Integer,NaN,serum cholestoral,mg/dl,no
5,fbs,Feature,Categorical,NaN,fasting blood sugar > 120 mg/dl,NaN,no
6,restecg,Feature,Categorical,NaN,NaN,NaN,no
7,thalach,Feature,Integer,NaN,maximum heart rate achieved,NaN,no
8,exang,Feature,Categorical,NaN,exercise induced angina,NaN,no
9,oldpeak,Feature,Integer,NaN,ST depression induced by exercise relative to ...,NaN,no


In [ ]:
# Categorical / numeric feature groups (based on the UCI data dictionary above)
categorical_columns = ["sex", "cp", "fbs", "restecg", "exang", "slope", "ca", "thal"]
numeric_columns = ["age", "trestbps", "chol", "thalach", "oldpeak"]
categorical_columns, numeric_columns

(['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'thal'],
 ['age', 'trestbps', 'chol', 'thalach', 'oldpeak'])

In [ ]:
for column in categorical_columns:
    print(
        f"{column}: "
        f"values={sorted(X[column].dropna().unique().tolist())}, "
        f"missing={X[column].isna().sum()}"
    )

sex: values=[0, 1], missing=0
cp: values=[1, 2, 3, 4], missing=0
fbs: values=[0, 1], missing=0
restecg: values=[0, 1, 2], missing=0
exang: values=[0, 1], missing=0
slope: values=[1, 2, 3], missing=0
ca: values=[0.0, 1.0, 2.0, 3.0], missing=4
thal: values=[3.0, 6.0, 7.0], missing=2


### 3. Data Cleaning

In [ ]:
# Create a copy for data cleaning
df = X.copy()

# Add the target variable
df["target"] = y["num"]

df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0,0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0,2
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0,1
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0,0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0,0


In [ ]:
# Check missing values in the working dataset
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

In [ ]:
# Check the distribution of missing-value columns
print("ca:")
print(df["ca"].value_counts(dropna=False).sort_index())

print("\nthal:")
print(df["thal"].value_counts(dropna=False).sort_index())

ca:
ca
0.0    176
1.0     65
2.0     38
3.0     20
NaN      4
Name: count, dtype: int64

thal:
thal
3.0    166
6.0     18
7.0    117
NaN      2
Name: count, dtype: int64


## Missing Values

The dataset contains missing values in two features:

- `ca`: 4 missing values
- `thal`: 2 missing values

All other features and the target variable contain no missing values.

These missing values will be handled during the preprocessing stage using a Scikit-learn Pipeline.

In [ ]:
# Check for duplicate rows
df.duplicated().sum()

np.int64(0)

In [ ]:
# Check for potentially invalid values
print("Age:", df["age"].min(), "-", df["age"].max())
print("Resting BP:", df["trestbps"].min(), "-", df["trestbps"].max())
print("Cholesterol:", df["chol"].min(), "-", df["chol"].max())
print("Max Heart Rate:", df["thalach"].min(), "-", df["thalach"].max())
print("Oldpeak:", df["oldpeak"].min(), "-", df["oldpeak"].max())

Age: 29 - 77
Resting BP: 94 - 200
Cholesterol: 126 - 564
Max Heart Rate: 71 - 202
Oldpeak: 0.0 - 6.2


In [ ]:
# Check categorical values
for column in categorical_columns:
    print(f"{column}: {sorted(df[column].dropna().unique())}")

sex: [np.int64(0), np.int64(1)]
cp: [np.int64(1), np.int64(2), np.int64(3), np.int64(4)]
fbs: [np.int64(0), np.int64(1)]
restecg: [np.int64(0), np.int64(1), np.int64(2)]
exang: [np.int64(0), np.int64(1)]
slope: [np.int64(1), np.int64(2), np.int64(3)]
ca: [np.float64(0.0), np.float64(1.0), np.float64(2.0), np.float64(3.0)]
thal: [np.float64(3.0), np.float64(6.0), np.float64(7.0)]


### 4. Handling Missing Values

`ca` and `thal` contain a small number of missing values. Rather than dropping or imputing them here, imputation is deferred to the `SimpleImputer` step inside the Scikit-learn `Pipeline` built in `04_feature_engineering_pipeline.ipynb`, so it is learned only from training data (avoiding data leakage) and stays fully reproducible.

In [ ]:
from sklearn.impute import SimpleImputer

# Create an imputer for categorical features
categorical_imputer = SimpleImputer(strategy="most_frequent")

# Check the most frequent values
for column in ["ca", "thal"]:
    most_frequent = df[column].mode()[0]
    print(f"{column}: {most_frequent}")

ca: 0.0
thal: 3.0


In [ ]:
# Final data quality check
print("Missing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

Missing values:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

Duplicate rows:
0


### 5. Save Cleaned Dataset

The cleaned dataset (duplicates/invalid values checked, missing values *not yet* imputed) is saved to `data/heart_disease_clean.csv` so the rest of the notebooks can load it directly and each run independently from top to bottom.

In [ ]:
import os

os.makedirs("../data", exist_ok=True)
df.to_csv("../data/heart_disease_clean.csv", index=False)
print("Saved:", df.shape)

Saved: (303, 14)
